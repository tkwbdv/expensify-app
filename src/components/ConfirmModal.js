import React from "react"
import Modal from "react-modal";

if (process.env.NODE_ENV !== 'test') Modal.setAppElement('#root'); // prevents error when testing

const ConfirmModal = ({ showModal, onCloseModal, onRemove }) => {
  return (
    <Modal
      isOpen={showModal}
      onRequestClose={onCloseModal}
      overlayClassName="overlay"
      className="modal"
      closeTimeoutMS={200}
      contentLabel="confirm Remove"
    >
      <h3 className="modal__title">Remove this expense?</h3>
      <div className="modal__button-container">
        <button className="button button--secondary button--confirm" onClick={onRemove}>Yes</button>
        <button className="button button--confirm" onClick={onCloseModal}>No</button>
      </div>
    </Modal>
  );
};

export { ConfirmModal as default };