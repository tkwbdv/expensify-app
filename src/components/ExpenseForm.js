import React, { useState } from "react";

import moment from "moment";

import "react-dates/initialize";
import { SingleDatePicker } from 'react-dates';

const ExpenseForm = ({ formData, expense = {} }) => {
  const [description, setDescription] = useState(expense.description || "");
  const [amount, setAmount] = useState((expense.amount / 100) || "");
  const [note, setNote] = useState(expense.note || "");
  const [createdAt, setCreatedAt] = useState(moment(expense.createdAt) || moment());
  const [calendarFocused, setCalendarFocused] = useState(false);
  const [error, setError] = useState("");

  const onAmountChange = (e) => {
    let amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      setAmount(amount);
    }
  };

  const onDescriptionChange = (e) => {
    const description = e.target.value;
    setDescription(description);
  };

  const onNoteChange = (e) => {
    const note = e.target.value;
    setNote(note);
  };

  const onDateChange = (createdAt) => {
    if (createdAt) {
      setCreatedAt(createdAt)
    }
  };

  const onFormSubmit = (e) => {
    e.preventDefault();
    if (!description || !amount) {
      setError("Please provide description and amount.");
    } else {
      setError("");
      formData({
        description,
        amount: +(parseFloat(amount) * 100).toFixed(0),
        createdAt: +createdAt,  // momentObject -> unix timestamp
        note
      });
    }
  };


  return (
    <div>
      {error && <p>{error}</p>}
      <form onSubmit={onFormSubmit}>
        <input
          placeholder="Description"
          autoFocus
          value={description}
          onChange={onDescriptionChange}
        />
        <input
          placeholder="Amount"
          type="text"
          value={amount}
          onChange={onAmountChange}
        />
        <SingleDatePicker
          date={createdAt}
          onDateChange={onDateChange}
          focused={calendarFocused}
          onFocusChange={({ focused }) => setCalendarFocused(focused)}
          id="your_unique_id"
          noBorder={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
          hideKeyboardShortcutsPanel={true}
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          value={note}
          onChange={onNoteChange}
        />
        <input
          type="submit"
          value={expense.description ? "Update Expense" : "Create Expense"}
        />
      </form>
    </div >
  )
};

export default ExpenseForm;