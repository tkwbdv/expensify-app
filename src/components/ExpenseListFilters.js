import React, { useState } from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from 'react-dates';

export const ExpenseListFilters = ({ filters, setStartDate, setEndDate, setTextFilter, sortByAmount, sortByDate }) => {
  const [focusedInput, setFocusedInput] = useState(null);

  const onDatesChange = ({ startDate, endDate }) => {
    setStartDate(startDate);
    setEndDate(endDate);
  };

  const onTextChange = (e) => {
    setTextFilter(e.target.value);
  }

  const onSortChange = (e) => {
    if (e.target.value === "amount") {
      sortByAmount();
    } else if (e.target.value === "date") {
      sortByDate();
    }
  }

  return (
    <div>
      <input
        type="text"
        value={filters.text}
        onChange={onTextChange}
      />
      <select
        value={filters.sortBy}
        onChange={onSortChange}
      >
        <option value="amount">Amount</option>
        <option value="date">Date</option>
      </select>
      <DateRangePicker
        startDate={filters.startDate}
        startDateId="your_unique_start_date_id"
        endDate={filters.endDate}
        endDateId="your_unique_end_date_id"
        onDatesChange={onDatesChange}
        focusedInput={focusedInput}
        onFocusChange={focusedInput => setFocusedInput(focusedInput)}
        hideKeyboardShortcutsPanel={true}
        noBorder={true}
        isOutsideRange={() => false}
        numberOfMonths={1}
        showClearDates={true}
      />
    </div>
  );
};


const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);