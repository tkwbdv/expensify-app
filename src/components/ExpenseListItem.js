import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>{amount} - {moment(createdAt).format("MMMM Do YYYY")}</p>
  </div>
);

export default ExpenseListItem;