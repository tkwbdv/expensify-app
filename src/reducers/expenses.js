// Expenses Reducer

const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(expense => expense.id !== action.id);
    case "EDIT_EXPENSE":
      return state.map(expense => {
        return expense.id === action.id
          ? { ...expense, ...action.updates }
          : expense;
      });
    default:
      return state;
  }
};

export default expensesReducer;