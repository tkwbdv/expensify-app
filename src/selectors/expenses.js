// Get visible expenses

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const startDateMatch = startDate ? expense.createdAt >= +startDate : true;
    const endDateMatch = endDate ? expense.createdAt <= +endDate : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === "amount") return a.amount < b.amount ? 1 : -1;
    if (sortBy === "date") return a.createdAt < b.createdAt ? 1 : -1;
    return 0;
  });
};

export default getVisibleExpenses;