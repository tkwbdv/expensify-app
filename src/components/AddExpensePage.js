import React from "react";
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export const AddExpensePage = ({ history, addExpense }) => (
  <div>
    <h1>Add Expense</h1>
    <ExpenseForm
      formData={(expense) => {
        addExpense(expense);
        history.push("/");
      }}
    />
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);