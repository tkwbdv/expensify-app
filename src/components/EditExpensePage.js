import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import NotFoundPage from "./NotFoundPage";

export const EditExpensePage = ({ match, startEditExpense, startRemoveExpense, expense, history }) => {
  if (!expense) return (
    <NotFoundPage />
  );

  const formData = (expenseData) => {
    startEditExpense(match.params.id, expenseData);
    history.push("/");
  }

  const onRemove = () => {
    startRemoveExpense({ id: match.params.id });
    history.push("/");
  };

  return (
    <div>
      <h1>Edit Expense</h1>
      <ExpenseForm
        formData={formData}
        expense={expense}
      />
      <button onClick={onRemove}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expenseData) => dispatch(startEditExpense(id, expenseData)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);