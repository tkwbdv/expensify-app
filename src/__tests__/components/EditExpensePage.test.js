import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let editExpense, removeExpense, match, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  match = { params: { id: expenses[2].id } };
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      match={match}
      editExpense={editExpense}
      removeExpense={removeExpense}
      expense={expenses[2]}
      history={history}
    />
  );
});

test("should render EditExpensePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should render NotFoundPage if expense id does not exist", () => {
  wrapper.setProps({ expense: undefined });
  expect(wrapper).toMatchSnapshot();
});

test("should handle editExpense", () => {
  wrapper.find("ExpenseForm").prop("formData")(expenses[2]);
  expect(editExpense).toHaveBeenLastCalledWith(match.params.id, expenses[2]);
  expect(history.push).toHaveBeenLastCalledWith("/");
});

test("should handle removeExpense", () => {
  wrapper.find("button").simulate("click");
  expect(removeExpense).toHaveBeenLastCalledWith({ id: match.params.id });
  expect(history.push).toHaveBeenLastCalledWith("/");
});