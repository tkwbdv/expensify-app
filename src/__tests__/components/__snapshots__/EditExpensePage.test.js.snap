// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": "1970-01-01T00:00:01.000Z",
        "description": "Credit Card",
        "id": 3,
        "note": "",
      }
    }
    formData={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`should render NotFoundPage if expense id does not exist 1`] = `<NotFoundPage />`;
